docs/expressions/exceptions.md

```diff
@@ -78,6 +78,46 @@ __Do I have to have an else error handler to have a then block?__ No. You can ha
 
 __Will my then block really always be executed, even if I return inside the try?__ Yes, your `then` expression will __always__ be executed when the try block is complete. The only way it won't be is if the try never completes (due to an infinite loop), the machine is powered off or the process is killed (and then, maybe).
 
+# With blocks
+
+A `with` expression can be used to ensure disposal of an object when it is no longer needed. A common case is a database connection which needs to be closed after use to avoid resource leaks on the server. For example:
+
+```pony
+with obj = SomeObjectThatNeedsDisposing() do
+  // use obj
+end
+```
+
+`obj.dispose()` will be called whether the code inside the `with` block completes successfully or raises an error. To take part in a `with` expression, the object that needs resource clean-up must therefore provide a `dispose()` method:
+
+```pony
+class SomeObjectThatNeedsDisposing
+  // constructor, other functions
+
+  fun dispose() =>
```

Comment:
This may or may not matter but as a small comment - it seems in most (if not all?) cases, cleaning up resources would require more than just read-only access to the object.

To avoid confusing readers, perhaps it would be most idiomatic for this to be `fun ref dispose`?